### PR TITLE
*: reduce runtime.growslice by preallocation

### PR DESCRIPTION
**Pull request by xiekeyi98**
Signed-off-by: Keyi Xie <xiekeyi98@gmail.com>

### What problem does this PR solve? 

Prealloc many slices to speed up.

### What is changed and how it works?
Prealloc what we can prealloc.

### Check List 

Tests 

 - Unit test
 - Integration test

```
xiekeyi@ubuntu:~/main$ cat a_test.go 
package main

import (
	"math/rand"
	"testing"
)

func BenchmarkIntPreAlloc(b *testing.B) {
	arrs := make([]int, 0, b.N)
	for i := 0; i < b.N; i++ {
		arrs = append(arrs, rand.Int())
	}
}

func BenchmarkIntNoneAlloc(b *testing.B) {
	var arrs []int
	for i := 0; i < b.N; i++ {
		arrs = append(arrs, rand.Int())
	}
}

func BenchmarkStructPreAlloc(b *testing.B) {
	arrs := make([]struct{}, 0, b.N)
	for i := 0; i < b.N; i++ {
		arrs = append(arrs, struct{}{})
	}
}

func BenchmarkStructNoneAlloc(b *testing.B) {
	var arrs []struct{}
	for i := 0; i < b.N; i++ {
		arrs = append(arrs, struct{}{})
	}
}

func BenchmarkInterfacePreAlloc(b *testing.B) {
	arrs := make([]interface{}, 0, b.N)
	for i := 0; i < b.N; i++ {
		arrs = append(arrs, nil)
	}
}

func BenchmarkInterfaceNoneAlloc(b *testing.B) {
	var arrs []interface{}
	for i := 0; i < b.N; i++ {
		arrs = append(arrs, nil)
	}
}

func BenchmarkStringPreAlloc(b *testing.B) {
	strs := make([]string, 0, b.N)
	for i := 0; i < b.N; i++ {
		strs = append(strs, "test")
	}
}

func BenchmarkStringNoneAlloc(b *testing.B) {
	var strs []string
	for i := 0; i < b.N; i++ {
		strs = append(strs, "test")
	}
}
func BenchmarkAllocStringLeakHalf(b *testing.B) {
	strs := make([]string, 0, b.N/2)
	for i := 0; i < b.N; i++ {
		strs = append(strs, "test")
	}
}
func BenchmarkAllocStringLeakMore(b *testing.B) {
	strs := make([]string, 0, b.N/3)
	for i := 0; i < b.N; i++ {
		strs = append(strs, "test")
	}
}
func BenchmarkAllocStringLeakMost(b *testing.B) {
	strs := make([]string, 0, b.N/100)
	for i := 0; i < b.N; i++ {
		strs = append(strs, "test")
	}
}

func BenchmarkAllocDouble(b *testing.B) {
	strs := make([]string, 0, b.N*2)
	for i := 0; i < b.N; i++ {
		strs = append(strs, "test")
	}
}
xiekeyi@ubuntu:~/main$ vim a_test.go 
xiekeyi@ubuntu:~/main$ go test -bench=. -benchmem
goos: linux
goarch: amd64
BenchmarkIntPreAlloc-4           	30000000	        49.9 ns/op	       8 B/op	       0 allocs/op
BenchmarkIntNoneAlloc-4          	20000000	        52.8 ns/op	      41 B/op	       0 allocs/op
BenchmarkStructPreAlloc-4        	2000000000	         0.49 ns/op	       0 B/op	       0 allocs/op
BenchmarkStructNoneAlloc-4       	300000000	         6.15 ns/op	       0 B/op	       0 allocs/op
BenchmarkInterfacePreAlloc-4     	100000000	        34.5 ns/op	      16 B/op	       0 allocs/op
BenchmarkInterfaceNoneAlloc-4    	20000000	        99.8 ns/op	      80 B/op	       0 allocs/op
BenchmarkStringPreAlloc-4        	50000000	        30.3 ns/op	      16 B/op	       0 allocs/op
BenchmarkStringNoneAlloc-4       	10000000	       133 ns/op	      82 B/op	       0 allocs/op
BenchmarkAllocStringLeakHalf-4   	20000000	        92.6 ns/op	      65 B/op	       0 allocs/op
BenchmarkAllocStringLeakMore-4   	20000000	        87.5 ns/op	      60 B/op	       0 allocs/op
BenchmarkAllocStringLeakMost-4   	10000000	       140 ns/op	      86 B/op	       0 allocs/op
BenchmarkAllocDouble-4           	fatal error: runtime: out of memory
```
Sorry to say my memory is not enough...

### CODE REVIEW

File: types/datum.go [hunk 1]
```diff
@@ -1794,7 +1794,7 @@ func handleTruncateError(sc *stmtctx.StatementContext) error {
 
 // DatumsToString converts several datums to formatted string.
 func DatumsToString(datums []Datum, handleSpecialValue bool) (string, error) {
-	var strs []string
+	strs := make([]string, .0, len(datums))
```
Comment (zz-jason) [inline types/datum.go]:
```suggestion
	strs := make([]string, 0, len(datums))
```

Comment (xiekeyi98) [reply types/datum.go]: Done.
Sorry for my carelessness.

**Pull request merged**